### ISSUE TITLE
wallet.keys

### ISSUE THREAD

**Issue by DangerousFreedom1984** [seraphis-migration/wallet3#33]
I would like to start a discussion about what should be stored in the wallet.keys file.
As explained by [this](https://monero.stackexchange.com/questions/2803/what-data-do-the-wallet-files-contain) post and you can find in the source code [here](https://github.com/monero-project/monero/blob/master/src/wallet/wallet2.cpp#L3887), the wallet.keys file does not contain only the corresponding keys to your corresponding wallet (view-only,normal,multisig). It has also a lot of other settings (like restoring_height, inactivity_lock_timeout, fee priority,...) that not necessarily is essential to load a wallet but make the user experience much more pleasant.

I see two scenarios here:

1. We continue with the philosophy of embedding things into the .keys file for jamtis/seraphis.
    1.  Have just one big class like it is today (for example wallet3) and insert all the different parameters of all the wallet types into this class, for example 'm_view_only' would be part of all wallets.
    2.   Create different classes of wallets each with its own particular parameters.

2. We split what is related to settings and what are private keys. 
    1.  Move all the settings to the 'wallet' file where the cache is. So normally we would have the 'wallet' and 'wallet.keys' files.
    2.  Create a new wallet.settings file where it would contain all the settings for the wallet. So normally we would have the 'wallet', 'wallet.settings' and 'wallet.keys' files.

I'm in favor of either the 1.ii or 2.ii scenarios as I believe that we could save space when storing things and we could keep things better organized. In the other hand, we would need to write more code too as each wallet tier will have its own classes for the .keys (and .settings files) although it could be easily managed by identifying the common parameters between them and using inheritance.
What do you guys think? Am I missing a scenario here?

Practically speaking, if we decide to go for splitting things, I would start making a list of common parameters of the wallets and start implementing a normal and view-only wallet with them so we can start discussing the structure of the wallet over something more concrete.

In any case, we will need to somehow copy a good deal of the old code (at least the loading part of old wallets) into the wallet3 classes (unless we don't entirely throw away wallet2 but I believe that this discussion is over) because we need to be able to deal with old wallets. The question would be how to store them again so it goes back to the beginning of the post.

**Comment by rbrunner7** [comment]
I think it's a good idea to separate the question of the design of runtime classes to organize and process data on the one side, and the question storing those things on the other side.

According to my experience, you can begin with almost completely ignoring the question of storing things into files and assume that later somehow things will store alright. Which they will, don't worry.

Concentrate on modelling classes with a good, logical, and appropriate design. Using inheritance where that's a good idea. Using objects that own a number of objects, either in a mandatory way (sub-object must exist) or in an optional way (sub-object may or may not exist, with pointer `null` in the latter case) where that's a good approach to solve a need. With good naming. And so on.

I am not sure, but maybe a good collection of classes to store all possible keys of all wallet types, and all supported options, will use a mix of inheritance and object-owning.

Just brainstorming now, right, you have to work out the details, but I would check whether something like a `view_only_wallet_keys` class makes sense, with a `full_wallet_keys` class derived from that, and finally a `multisig_wallet_keys` again derived from that. Or even more levels, because Jamtis.

General options could go into a separate class of their own. Or maybe several classes, if it turns out that not all general options make sense for all 3 types of wallets. Or maybe not, because the set of options will be rather modest anyway, and easy to handle nevertheless.

A top level `settings` class could then own one of the mentioned "keys" classes and one "general options" object.

Deciding how to actually store those things into files, and how to distribute them between several files or write them all into a single file, is much easier if you have everything neatly defined right in front of you.

**Comment by rbrunner7** [comment]
Anyway, anticipating that "how many files" decision a bit:

Now you have two files, one that is "holy" and "valuable" and that you should better not lose, otherwise you have to restore from seed, and one that you can throw away without losing too much because the system is able to re-create it with the help from data in the first file, and data from the blockchain.

I think this arrangement stood the test of time, and right now I don't see any good reason to do something differently in this regard. One single file is certainly not a better idea, and neither is adding more files IMO.

**Comment by rbrunner7** [comment]
> In any case, we will need to somehow copy a good deal of the old code (at least the loading part of old wallets) into the wallet3 classes

I think we are talking about the code to load .keys file here. Yes, I think we must be able to load all the data in these files, to offer a migration path, and to be able to do so without a need to keep the complete massive `wallet2` class around.

I was thinking maybe it would make sense to refactor that loading code out of `wallet2` into a separate class, so our new wallet would be able to use that as well. But right now I am not a fan anymore of this idea, with the following argument: Invest the absolute mimimum of work and time into `wallet2` that we can get away with, along the lines of "It's dead, Jim".

If simply ripping out code / duplicating code from `wallet2` is faster than refactoring out a class properly, which it probably is, I would recommend that.

**Comment by UkoeHB** [comment]
> Just brainstorming now, right, you have to work out the details, but I would check whether something like a view_only_wallet_keys class makes sense, with a full_wallet_keys class derived from that, and finally a multisig_wallet_keys again derived from that. Or even more levels, because Jamtis.

For composition I really recommend explicitly composing instead of using inheritance to piece things together. You get the same result but it's a lot easier to reason about the shape of an object when the nesting direction is 'down' instead of 'up'. It also ensures slicing into the sub-parts is explicit instead of implicit, which aids call-site clarity. Otherwise you have to continually wonder 'am I being sliced here?'. Moreover, inheritance means your base class is tied in some sense to all inheritors, even though when adding a component into your new object, the component is expected to be an independent self-contained thing. Inheritance should be mostly limited to lifting interfaces away from implementations, with only edge cases otherwise.

**Comment by rbrunner7** [comment]
Yeah, whole generations of IT pros and theorist have locked horns and banged heads over such questions :)

What is probably uncontroversial is that often things are pressed into inheritance that don't belong there.

If a full wallet is not a view-only wallet a bit "more clever" or with a bit more "bells and whistles", and a multisig wallet is not an even more clever view-only wallet, in any reasonable sense, then it's not a hierarchy, and inheritance is not a good idea, agreed.
